pkg/pullsync/pullstorage/pullstorage.go

```diff
@@ -38,88 +41,126 @@ type Storer interface {
 	Set(ctx context.Context, mode storage.ModeSet, addrs ...swarm.Address) error
 	// Has chunks.
 	Has(ctx context.Context, addr swarm.Address) (bool, error)
+
+	io.Closer
 }
 
-// ps wraps storage.Storer.
-type ps struct {
+// PullStorer wraps storage.Storer.
+type PullStorer struct {
 	storage.Storer
+	intervalFlight singleflight.Group
+	metrics        metrics
+	psCtx          context.Context
+	psCancel       context.CancelFunc
 }
 
 // New returns a new pullstorage Storer instance.
-func New(storer storage.Storer) Storer {
-	return &ps{
-		Storer: storer,
+func New(storer storage.Storer) *PullStorer {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &PullStorer{
+		Storer:   storer,
+		metrics:  newMetrics(),
+		psCtx:    ctx,
+		psCancel: cancel,
 	}
 }
 
-// IntervalChunks collects chunk for a requested interval.
-func (s *ps) IntervalChunks(ctx context.Context, bin uint8, from, to uint64, limit int) (chs []swarm.Address, topmost uint64, err error) {
-	// call iterator, iterate either until upper bound or limit reached
-	// return addresses, topmost is the topmost bin ID
-	var (
-		timer  *time.Timer
-		timerC <-chan time.Time
-	)
-	ch, dbClosed, stop := s.SubscribePull(ctx, bin, from, to)
-	defer func(start time.Time) {
-		stop()
-		if timer != nil {
-			timer.Stop()
-		}
-	}(time.Now())
-
-	var nomore bool
+type intervalFlightResult struct {
+	chs     []swarm.Address
+	topmost uint64
+}
 
-LOOP:
-	for limit > 0 {
-		select {
-		case v, ok := <-ch:
-			if !ok {
-				nomore = true
-				break LOOP
-			}
-			chs = append(chs, v.Address)
-			if v.BinID > topmost {
-				topmost = v.BinID
+// IntervalChunks collects chunk for a requested interval.
+func (s *PullStorer) IntervalChunks(ctx context.Context, bin uint8, from, to uint64, limit int) ([]swarm.Address, uint64, error) {
+	s.metrics.TotalSubscribePullRequests.Inc()
+	defer s.metrics.TotalSubscribePullRequestsComplete.Inc()
+	resultC := s.intervalFlight.DoChan(subKey(bin, from, to, limit), func() (interface{}, error) {
```

Comment:
I believe that the most correct thing to do in regards to cancelation of this singleflight is to be context aware of all callers' contexts. I did an implementation of such singleflight a few years ago https://pkg.go.dev/resenje.org/singleflight#Group.Do. It lacks the DoChan, but I think that it is not needed in this case, as context are handled internally in that package.

The problem that I see is when the passeed context is done, the IntervalChunks function returns, but the created goroutine is still waiting until the PullStorer instance is closed. This should result with the goroutine leak, and the suggested approach would solve this problem.

Reply:
@aloknerurkar @acud I have created an example of the described approach in this PR https://github.com/ethersphere/bee/pull/1891.

Reply:
Went through the change in your lib. This solution is definitely better. I will close this PR and we can use your changes.

So I wanted to run this by you anyway,

```
The problem that I see is when the passed context is done, the IntervalChunks function returns, but the created goroutine is still waiting until the PullStorer instance is closed. This should result with the goroutine leak, and the suggested approach would solve this problem.
```
So this was known. However, singleFlight writes the result to a buffered channel. Even if all the callers go away, the singleFlight request will eventually finish and exit. This is why I think there won't be any leaks here.

Also, I was thinking that if all the callers are canceled, pullsync will probably reissue the Subscribe with the same arguments. At which point, we will reuse any singleFlights that were still waiting.